### ISSUE TITLE
Regression introduced once using Sqeleton when using Python SDK

### ISSUE THREAD

**Issue by MattDelac** [datafold/data-diff#475]
## Describe the bug
I was using the python SDK of data-diff 0.2.8 and life was good
> data-diff = {extras = ["mysql", "postgresql"], version = "^0.2.8"}

Two days ago I updated data-diff to its latest version (0.5.0), and my pipelines stopped working.
The most common error was the following one
> TypeError: Incompatible types for column 'manually_marked_reliable':  Integer(precision=0, python_type=<class 'int'>) <-> UnknownColType(text='smallint')

Those pipelines compare tables between MySQL <> Redshift.

## What I tried
I could not find the root cause of the bug so I downgraded `data-diff` to a version until it worked.
Here is the summary
- data-diff 0.6.0 with sqeleton 0.0.8 ==> KO same error
- data-diff 0.5.0 with sqeleton 0.0.6 ==> KO same error
- data-diff 0.4.1 with sqeleton 0.0.5 ==> KO same error
- data-diff 0.3.2 with sqeleton 0.0.5 ==> KO same error
- data-diff 0.3.0 (sqeleton got removed) ==> OK it works
- data-diff 0.2.8 (sqeleton did not exist at that time) ==> OK it works

Please note that I ran the exact same pipeline with the exact same config/parameters when testing the different versions of `data-diff`

More info on Slack in this [public thread](https://locallyoptimistic.slack.com/archives/C03HUNGQV0S/p1680559049227069)

## Problem with the SDK only
Interestingly, the behavior is different if I use the CLI vs the Python SDK.
This is my latest test using  the following dependencies
```sh
Name: sqeleton
Version: 0.0.8
Required-by: data-diff
---
Name: data-diff
Version: 0.6.0
Required-by: pipo
```

### Using CLI -> OK
```sh
poetry run data-diff mysql:xxx YYY redshift:xxx/ YYY -t updated_at --min-age 30min --threads 12 --stats
> ...
1036255 rows in table A
1036255 rows in table B
0 rows exclusive to table A (not present in B)
0 rows exclusive to table B (not present in A)
0 rows updated
1036255 rows unchanged
0.00% difference score
```

### Using Python SDK with the following logic -> NOT OK
```python
from data_diff import connect_to_table, diff_tables
from data_diff.table_segment import TableSegment

def compute_diff(
    src: TableSegment,
    dst: TableSegment,
    table: TableConfig, # Custom data model
    extra_columns: Tuple[str],
    unique_id: str,
) -> Dict[str, Any]:
    src_row_number: int = src.count()
    dst_row_number: int = dst.count()
    bisection_factor = A_NUMBER_CALCULATED

    result: Dict[str, Any] = {
        "dst_row_count": dst_row_number,
        "unique_id": unique_id,
    }

    start_datetime = datetime.utcnow()
    diff_iter = diff_tables(table1=src, table2=dst, bisection_factor=bisection_factor)
    stats = diff_iter.get_stats_dict()

    logger.info(f"Stats from the diff_iter {stats}")

    result["minus"] = stats["exclusive_A"]
    result["plus"] = stats["exclusive_B"]
    result["mismatched"] = stats["updated"]

    result.update(
        {
            "time_to_diff_in_seconds": (datetime.utcnow() - start_datetime).seconds,
        }
    )

    logger.info(f"Result: {result}")
    return result
```

It fails at that line `stats = diff_iter.get_stats_dict()` with the following error
```sh
└── 11:52:47 | ERROR   | Task 'compute_diff': Exception encountered during task execution!
Traceback (most recent call last):
  File "/Users/mattdelacour/Documents/src/coinlist/pipo/.venv/lib/python3.10/site-packages/prefect/engine/task_runner.py", line 880, in get_task_run_state
    value = prefect.utilities.executors.run_task_with_timeout(
  File "/Users/mattdelacour/Documents/src/coinlist/pipo/.venv/lib/python3.10/site-packages/prefect/utilities/executors.py", line 468, in run_task_with_timeout
    return task.run(*args, **kwargs)  # type: ignore
  File "accuracy/check_accuracy.py", line 254, in compute_diff
    stats = diff_iter.get_stats_dict()
  File "/Users/mattdelacour/Documents/src/coinlist/pipo/.venv/lib/python3.10/site-packages/data_diff/diff_tables.py", line 169, in get_stats_dict
    diff_stats = self._get_stats()
  File "/Users/mattdelacour/Documents/src/coinlist/pipo/.venv/lib/python3.10/site-packages/data_diff/diff_tables.py", line 100, in _get_stats
    list(self)  # Consume the iterator into result_list, if we haven't already
  File "/Users/mattdelacour/Documents/src/coinlist/pipo/.venv/lib/python3.10/site-packages/data_diff/diff_tables.py", line 95, in __iter__
    for i in self.diff:
  File "/Users/mattdelacour/Documents/src/coinlist/pipo/.venv/lib/python3.10/site-packages/data_diff/diff_tables.py", line 246, in _diff_tables_wrapper
    raise error
  File "/Users/mattdelacour/Documents/src/coinlist/pipo/.venv/lib/python3.10/site-packages/data_diff/diff_tables.py", line 217, in _diff_tables_wrapper
    self._validate_and_adjust_columns(table1, table2)
  File "/Users/mattdelacour/Documents/src/coinlist/pipo/.venv/lib/python3.10/site-packages/data_diff/hashdiff_tables.py", line 99, in _validate_and_adjust_columns
    raise TypeError(f"Incompatible types for column '{c1}':  {col1} <-> {col2}")
TypeError: Incompatible types for column 'converted_from_additional_allocation':  Integer(precision=0, python_type=<class 'int'>) <-> UnknownColType(text='smallint')
└── 11:52:47 | INFO    | Task 'compute_diff': Finished task run for task with final state: 'Failed'
└── 11:52:47 | INFO    | Task 'save_general_stats_in_redshift': Starting task run...
└── 11:52:47 | INFO    | Task 'save_general_stats_in_redshift': Finished task run for task with final state: 'TriggerFailed'
└── 11:52:47 | INFO    | Task 'save_detailled_errors_in_redshift': Starting task run...
└── 11:52:47 | INFO    | Task 'save_detailled_errors_in_redshift': Finished task run for task with final state: 'TriggerFailed'
└── 11:52:47 | INFO    | Flow run FAILED: some reference tasks failed.
```

**Describe the environment**
MacOS. code running within poetry virtual env. Same behavior when running on my Linux servers.
Python 3.10

**Comment by MattDelac** [comment]
Note that I use the same logic as in the CLI
https://github.com/datafold/data-diff/blob/v0.6.0/data_diff/__main__.py#L446-L456

**Comment by MattDelac** [comment]
Even though a patch has been pushed in sqeleton https://github.com/datafold/sqeleton/pull/37, we should make sure there is consistency between the CLI and the Python SDK (either through easier helper methods and/or documentation)

**Comment by williebsweet** [comment]
Hey @MattDelac, since the subject of this issue was resolved by [datafold/sqeleton#37](https://github.com/datafold/sqeleton/pull/37), can we close out this issue?

I agreed that we need further alignment between the CLI and Python SDK, but it would be better to open up a new issue for each specific case.

**Comment by dlawin** [comment]
Closing as we haven't seen a response. Feel free to reopen
